File: rs/pragmastat/tests/assume_sorted_tests.rs
//! Direct coverage for the raw API's `assume_sorted = true` branch.
//!
//! The dual-path reference tests only ever call the raw estimators with
//! `assume_sorted = false`; the `= true` branch is reached only transitively via
//! `Sample` (which passes its cached sorted view). This file exercises the
//! `= true` branch directly.
//!
//! For ORDER-INDEPENDENT estimators, sorting the input ascending and calling
//! with `assume_sorted = true` must equal the call on the unsorted input with
//! `assume_sorted = false`.
//!
//! For SHUFFLE-based bounds, `assume_sorted` only swaps an internal re-sort for
//! the supplied pre-sorted view; it never affects the shuffle (which always runs
//! on the ORIGINAL order). So on a genuinely SORTED slice with a fixed seed the
//! result must be byte-identical for `true` vs `false`.
//!
//! Every comparison here is BIT-EXACT on the binary64 payload, not approximate.
//! Both sides are the same kernel fed the same values in the same order; the
//! flag only decides whether that order is re-established by an internal sort or
//! taken on trust. `assume_sorted = false` sorts a copy with `total_cmp`, which
//! yields exactly the slice the `= true` leg is handed, so there is no
//! arithmetic anywhere between the two paths for a rounding to enter. A
//! tolerance would not describe a numerical fact, it would only hide a defect.
//! The `ratio` legs go through log/exp, but BOTH legs go through the same
//! log/exp on the same values, so they are exact for the same reason.

mod conformance;

// Routed through the crate's shared `assert_bits_eq` rather than a private
// comparison so that "exact" means the same thing here as in the fixture suites.
// It compares the raw payloads rather than `==`, which has blind spots:
// `-0.0 == 0.0` holds and `NaN == NaN` does not, and either would misreport a
// genuine divergence between the two paths.
use conformance::assert_bits_eq;
use pragmastat::estimators::raw;

const MISRATE: f64 = 0.3;
const SEED: &str = "pragmastat";

fn sorted_copy(x: &[f64]) -> Vec<f64> {
    let mut v = x.to_vec();
    v.sort_unstable_by(|a, b| a.total_cmp(b));
    v
}

fn unsorted_x() -> Vec<f64> {
    vec![3.0, 1.0, 2.0, 5.0, 4.0, 8.0, 6.0, 7.0]
}

fn unsorted_y() -> Vec<f64> {
    vec![9.0, 11.0, 10.0, 13.0, 12.0, 16.0, 14.0, 15.0]
}

// --- Order-independent scalar estimators ---

#[test]
fn center_sorted_true_equals_unsorted_false() {
    let x = unsorted_x();
    let sorted = sorted_copy(&x);
    let want = raw::center(&x, false).unwrap();
    let got = raw::center(&sorted, true).unwrap();
    assert_bits_eq("center", got, want);
}

#[test]
fn spread_sorted_true_equals_unsorted_false() {
    let x = unsorted_x();
    let sorted = sorted_copy(&x);
    let want = raw::spread(&x, false).unwrap();
    let got = raw::spread(&sorted, true).unwrap();
    assert_bits_eq("spread", got, want);
}

#[test]
fn shift_sorted_true_equals_unsorted_false() {
    let x = unsorted_x();
    let y = unsorted_y();
    let want = raw::shift(&x, &y, false).unwrap();
    let got = raw::shift(&sorted_copy(&x), &sorted_copy(&y), true).unwrap();
    assert_bits_eq("shift", got, want);
}

#[test]
fn ratio_sorted_true_equals_unsorted_false() {
    let x = unsorted_x();
    let y = unsorted_y();
    let want = raw::ratio(&x, &y, false).unwrap();
    let got = raw::ratio(&sorted_copy(&x), &sorted_copy(&y), true).unwrap();
    assert_bits_eq("ratio", got, want);
}

#[test]
fn disparity_sorted_true_equals_unsorted_false() {
    let x = unsorted_x();
    let y = unsorted_y();
    let want = raw::disparity(&x, &y, false).unwrap();
    let got = raw::disparity(&sorted_copy(&x), &sorted_copy(&y), true).unwrap();
    assert_bits_eq("disparity", got, want);
}

// --- Order-independent bounds estimators ---

#[test]
fn center_bounds_sorted_true_equals_unsorted_false() {
    let x = unsorted_x();
    let want = raw::center_bounds(&x, MISRATE, false).unwrap();
    let got = raw::center_bounds(&sorted_copy(&x), MISRATE, true).unwrap();
    assert_bits_eq("center_bounds lower", got.lower, want.lower);
    assert_bits_eq("center_bounds upper", got.upper, want.upper);
}

#[test]
fn shift_bounds_sorted_true_equals_unsorted_false() {
    let x = unsorted_x();
    let y = unsorted_y();
    let want = raw::shift_bounds(&x, &y, MISRATE, false).unwrap();
    let got = raw::shift_bounds(&sorted_copy(&x), &sorted_copy(&y), MISRATE, true).unwrap();
    assert_bits_eq("shift_bounds lower", got.lower, want.lower);
    assert_bits_eq("shift_bounds upper", got.upper, want.upper);
}

#[test]
fn ratio_bounds_sorted_true_equals_unsorted_false() {
    let x = unsorted_x();
    let y = unsorted_y();
    let want = raw::ratio_bounds(&x, &y, MISRATE, false).unwrap();
    let got = raw::ratio_bounds(&sorted_copy(&x), &sorted_copy(&y), MISRATE, true).unwrap();
    // Both legs log-transform, run the same shift bounds, and exponentiate back;
    // sorting a positive sample and log-transforming it produces exactly the
    // array the other leg sorts, because log is monotonic. Same route, so exact.
    assert_bits_eq("ratio_bounds lower", got.lower, want.lower);
    assert_bits_eq("ratio_bounds upper", got.upper, want.upper);
}

// --- Shuffle-based bounds: identical on a SORTED slice with a fixed seed ---
//
// On a genuinely sorted slice the shuffle order is identical for both calls and
// the only difference is whether the (valid) sorted view is reused, so the
// result must be byte-identical.

#[test]
fn spread_bounds_sorted_true_equals_false_byte_identical() {
    let sorted = sorted_copy(&unsorted_x());
    let want = raw::spread_bounds_with_seed(&sorted, MISRATE, SEED, false).unwrap();
    let got = raw::spread_bounds_with_seed(&sorted, MISRATE, SEED, true).unwrap();
    assert_bits_eq("spread_bounds lower", got.lower, want.lower);
    assert_bits_eq("spread_bounds upper", got.upper, want.upper);
}

#[test]
fn disparity_bounds_sorted_true_equals_false_byte_identical() {
    let sorted_x = sorted_copy(&unsorted_x());
    let sorted_y = sorted_copy(&unsorted_y());
    let want = raw::disparity_bounds_with_seed(&sorted_x, &sorted_y, MISRATE, SEED, false).unwrap();
    let got = raw::disparity_bounds_with_seed(&sorted_x, &sorted_y, MISRATE, SEED, true).unwrap();
    assert_bits_eq("disparity_bounds lower", got.lower, want.lower);
    assert_bits_eq("disparity_bounds upper", got.upper, want.upper);
}

// NOTE: There is deliberately NO "spread_bounds inert on UNSORTED input" test.
// On unsorted input with assume_sorted = true the sparity check feeds unsorted
// data to the sorted-only spread_impl kernel, which is UNDEFINED BEHAVIOR: it
// may trip the convergence guard (iteration cap or stall detection) and error,
// or pass by luck for a particular input. The only well-defined inertness is on
// a SORTED slice (covered by spread_bounds_sorted_true_equals_false_byte_identical
// above).

// --- n==2 center midpoint order-symmetry ---
//
// With assume_sorted = true the midpoint sees the RAW order (the normalizing
// sort would hide any asymmetry). The 0.5*a + 0.5*b form is exactly symmetric
// in operand order, so reversing the two inputs must yield a BIT-IDENTICAL
// result. An asymmetric midpoint such as a + (b - a) * 0.5 would return
// -3.4000000000000004 for the reversed order below, so this guards an EXACT
// equality, not an approximate one.

#[test]
fn center_n2_midpoint_is_order_symmetric() {
    let forward = raw::center(&[-5.0, -1.8], true).unwrap();
    let reversed = raw::center(&[-1.8, -5.0], true).unwrap();
    assert_bits_eq("center n=2 forward", forward, -3.4);
    assert_bits_eq("center n=2 reversed", reversed, -3.4);
    assert_bits_eq("center n=2 order symmetry", reversed, forward);
}

File: cs/Pragmastat.Tests/Estimators/AssumeSortedTests.cs
namespace Pragmastat.Tests.Estimators;

/// <summary>
/// Directly exercises the raw (native-array) API's assumeSorted=true branch. The
/// dual-path reference tests only ever pass assumeSorted=false; the =true branch is
/// reached only transitively via Sample. This adds a direct check.
///
/// For ORDER-INDEPENDENT estimators, sorting the input ascending and calling with
/// assumeSorted=true must equal the call on the unsorted input with assumeSorted=false.
///
/// For SHUFFLE-based bounds, assumeSorted only affects the internal sparity check, never
/// the shuffle, so on an already-SORTED array with a fixed seed the result must be
/// byte-identical for true vs false.
///
/// Every comparison here is BITWISE, not tolerant. Both sides are the same kernel on the
/// same values, reached by two routes: assumeSorted=false sorts a copy and hands the kernel
/// exactly the array assumeSorted=true hands it directly. No arithmetic separates the two,
/// so there is nothing for a rounding difference to come from. Either the kernel selects the
/// same value and the payloads match to the last bit, or one route does something the other
/// does not, which is a defect. An epsilon there measures nothing and hides that defect.
/// Ratio and RatioBounds keep the exact comparison for the same reason: both sides take the
/// same log -> shift -> exp route (log is monotonic, so sorting before or after it produces
/// the identical array), so the transcendental calls see identical arguments.
/// </summary>
public class AssumeSortedTests
{
  private const string Seed = "pragmastat";
  private static readonly Probability Misrate = new(0.3);

  private static readonly double[] X = { 3, 1, 2, 5, 4, 8, 6, 7 };
  private static readonly double[] Y = { 9, 11, 10, 13, 12, 16, 14, 15 };

  private static double[] Sorted(double[] values)
  {
    var copy = (double[])values.Clone();
    Array.Sort(copy);
    return copy;
  }

  private static void AssertSameBounds(string context, Bounds want, Bounds got)
  {
    BitwiseAssert.Equal(want.Lower, got.Lower, $"{context}, lower");
    BitwiseAssert.Equal(want.Upper, got.Upper, $"{context}, upper");
  }

  // --- Order-independent scalar estimators ---

  [Fact]
  public void Center_AssumeSortedRoundtrip()
  {
    double want = Toolkit.Center(X, assumeSorted: false);
    double got = Toolkit.Center(Sorted(X), assumeSorted: true);
    BitwiseAssert.Equal(want, got, "Center(x, assumeSorted: false) vs Center(sorted x, assumeSorted: true)");
  }

  [Fact]
  public void Spread_AssumeSortedRoundtrip()
  {
    double want = Toolkit.Spread(X, assumeSorted: false);
    double got = Toolkit.Spread(Sorted(X), assumeSorted: true);
    BitwiseAssert.Equal(want, got, "Spread(x, assumeSorted: false) vs Spread(sorted x, assumeSorted: true)");
  }

  [Fact]
  public void Shift_AssumeSortedRoundtrip()
  {
    double want = Toolkit.Shift(X, Y, assumeSorted: false);
    double got = Toolkit.Shift(Sorted(X), Sorted(Y), assumeSorted: true);
    BitwiseAssert.Equal(want, got,
      "Shift(x, y, assumeSorted: false) vs Shift(sorted x, sorted y, assumeSorted: true)");
  }

  [Fact]
  public void Ratio_AssumeSortedRoundtrip()
  {
    double want = Toolkit.Ratio(X, Y, assumeSorted: false);
    double got = Toolkit.Ratio(Sorted(X), Sorted(Y), assumeSorted: true);
    BitwiseAssert.Equal(want, got,
      "Ratio(x, y, assumeSorted: false) vs Ratio(sorted x, sorted y, assumeSorted: true)");
  }

  [Fact]
  public void Disparity_AssumeSortedRoundtrip()
  {
    double want = Toolkit.Disparity(X, Y, assumeSorted: false);
    double got = Toolkit.Disparity(Sorted(X), Sorted(Y), assumeSorted: true);
    BitwiseAssert.Equal(want, got,
      "Disparity(x, y, assumeSorted: false) vs Disparity(sorted x, sorted y, assumeSorted: true)");
  }

  // --- Order-independent bounds estimators ---

  [Fact]
  public void CenterBounds_AssumeSortedRoundtrip()
  {
    var want = Toolkit.CenterBounds(X, Misrate, assumeSorted: false);
    var got = Toolkit.CenterBounds(Sorted(X), Misrate, assumeSorted: true);
    AssertSameBounds("CenterBounds(x, assumeSorted: false) vs CenterBounds(sorted x, assumeSorted: true)", want, got);
  }

  [Fact]
  public void ShiftBounds_AssumeSortedRoundtrip()
  {
    var want = Toolkit.ShiftBounds(X, Y, Misrate, assumeSorted: false);
    var got = Toolkit.ShiftBounds(Sorted(X), Sorted(Y), Misrate, assumeSorted: true);
    AssertSameBounds(
      "ShiftBounds(x, y, assumeSorted: false) vs ShiftBounds(sorted x, sorted y, assumeSorted: true)", want, got);
  }

  [Fact]
  public void RatioBounds_AssumeSortedRoundtrip()
  {
    var want = Toolkit.RatioBounds(X, Y, Misrate, assumeSorted: false);
    var got = Toolkit.RatioBounds(Sorted(X), Sorted(Y), Misrate, assumeSorted: true);
    AssertSameBounds(
      "RatioBounds(x, y, assumeSorted: false) vs RatioBounds(sorted x, sorted y, assumeSorted: true)", want, got);
  }

  // --- Shuffle-based bounds: assumeSorted never changes the shuffle (which always runs on
  // the original order). On a genuinely SORTED array with a fixed seed, the shuffle order is
  // identical for both calls and the only difference is whether the (valid) sorted view is
  // reused, so the result must be byte-identical. ---

  [Fact]
  public void SpreadBounds_AssumeSortedIdenticalOnSorted()
  {
    var sortedX = Sorted(X);
    var want = Toolkit.SpreadBounds(sortedX, Misrate, Seed, assumeSorted: false);
    var got = Toolkit.SpreadBounds(sortedX, Misrate, Seed, assumeSorted: true);
    AssertSameBounds("SpreadBounds(sorted x, assumeSorted: false vs true)", want, got);
  }

  [Fact]
  public void DisparityBounds_AssumeSortedIdenticalOnSorted()
  {
    var sortedX = Sorted(X);
    var sortedY = Sorted(Y);
    var want = Toolkit.DisparityBounds(sortedX, sortedY, Misrate, Seed, assumeSorted: false);
    var got = Toolkit.DisparityBounds(sortedX, sortedY, Misrate, Seed, assumeSorted: true);
    AssertSameBounds("DisparityBounds(sorted x, sorted y, assumeSorted: false vs true)", want, got);
  }

  // --- spreadBounds assumeSorted contract. assumeSorted=true is INERT only on SORTED input
  // (see SpreadBounds_AssumeSortedIdenticalOnSorted above): only the shuffle part is genuinely
  // order-independent. On UNSORTED input, assumeSorted=true is UNDEFINED BEHAVIOR, exactly like
  // every other estimator: the sparity (spread>0) check runs SpreadImpl(x, assumeSorted), and on
  // unsorted x with assumeSorted=true that feeds unsorted data to a sorted-only kernel (may hit
  // the iteration cap and ERROR, or pass by luck). So there is NO inert-on-UNSORTED test here. ---
}
